File: lint-arwaky/archive/infrastructure/javascript_call_tracer.py
"""JSTracer — Semantic analysis adapter for JavaScript/TypeScript files."""

from __future__ import annotations

import logging
import os
import re
import glob


from ..taxonomy import (
    CallChainList,
    Count,
    DataFlowList,
    DirectoryPath,
    FilePath,
    LineNumber,
    ResponseData,
    ResponseDataList,
    ScopeRef,
    SymbolName,
    SymbolNameList,
)
from ..contract import ISemanticTracerPort


logger = logging.getLogger("infrastructure.javascript_tracer")


# --- Inlined from naming_variant_generator (avoid cross-layer dependency
# capabilities→infrastructure) ---
class JSTracer(ISemanticTracerPort):
    """Regex-based semantic tracer for JavaScript/TypeScript files."""

    @staticmethod
    def _get_variant_dict(name):
        """Produce common naming variants for a symbol name."""
        n = name
        words = re.findall(r"[A-Za-z][a-z0-9]*|[A-Z]+(?=[A-Z][a-z0-9]|\b)|[0-9]+", n)
        words = [w.lower() for w in words]
        if not words:
            return {
                "snake_case": n,
                "camel_case": n,
                "pascal_case": n,
                "screaming_snake": n,
            }
        snake_case = "_".join(words)
        first = words[0]
        rest = "".join(w.capitalize() for w in words[1:])
        return {
            "snake_case": snake_case,
            "camel_case": first + rest,
            "pascal_case": "".join(w.capitalize() for w in words),
            "screaming_snake": snake_case.upper(),
        }

    @staticmethod
    def _build_variants_raw(name):
        """Build list of naming variants."""
        n = name
        d = JSTracer._get_variant_dict(n)
        kebab = d["snake_case"].replace("_", "-")
        return list(
            {
                n,
                d["snake_case"],
                d["camel_case"],
                d["pascal_case"],
                d["screaming_snake"],
                kebab,
            }
        )

    @staticmethod
    def _show_enclosing_scope(file_path, line):
        """Stub: return enclosing scope name for a given line in a JS/TS file."""
        return None

    @staticmethod
    def _find_flow(file_path: FilePath, var_name: SymbolName, start: LineNumber):
        """Stub: return data-flow entries for a variable in a JS/TS file."""
        return []

    def get_variant_dict(self, name: SymbolName) -> ResponseData:
        return ResponseData(value=self._get_variant_dict(str(name)))

    def build_variants(self, name: SymbolName) -> SymbolNameList:
        variants = self._build_variants_raw(str(name))
        return SymbolNameList(values=[SymbolName(value=v) for v in variants])

    def get_enclosing_scope(
        self, file_path: FilePath, line: LineNumber
    ) -> ScopeRef | None:
        result = self._show_enclosing_scope(str(file_path), int(line))
        if result:
            return ScopeRef(name=result)
        return None

    def get_symbol_locations(
        self, file_path: FilePath, symbol_name: SymbolName
    ) -> ResponseDataList:
        """Return empty list — not yet implemented for JS tracer."""
        return ResponseDataList(values=[])

    def find_flow(
        self,
        file_path: FilePath,
        var_name: SymbolName,
        start_line: LineNumber | None = None,
    ) -> DataFlowList:
        start = start_line if start_line is not None else LineNumber(value=0)
        flows = self._find_flow(
            FilePath(value=str(file_path)), SymbolName(value=str(var_name)), start
        )
        return DataFlowList(values=flows)

    def trace_call_chain(
        self, root_dir: DirectoryPath, target_name: SymbolName
    ) -> CallChainList:
        callers = []
        name = str(target_name)
        root = str(root_dir)
        call_pattern = re.compile(rf"\b{re.escape(name)}\s*\(")
        def_pattern = re.compile(rf"(?:function|class)\s+{re.escape(name)}\b")
        js_files = []
        for ext in ("*.js", "*.jsx", "*.ts", "*.tsx", "*.mjs"):
            js_files.extend(glob.glob(os.path.join(root, "**", ext), recursive=True))
        for filepath in js_files:
            try:
                with open(filepath, "r", encoding="utf-8") as f:
                    file_lines = f.readlines()
            except OSError:
                continue
            for i, line in enumerate(file_lines):
                if call_pattern.search(line) and not def_pattern.search(line):
                    rel_path = os.path.relpath(filepath, root)
                    callers.append(f"{rel_path}:{i + 1} -> {line.strip()}")
        return CallChainList(values=callers)

    def project_wide_rename(
        self, root_dir: DirectoryPath, old_name: SymbolName, new_name: SymbolName
    ) -> Count:
        root = str(root_dir)
        old = str(old_name)
        new = str(new_name)
        pattern = re.compile(
            rf"""
            (
                `(?:\\.|[^`\\])*`             |
                \"(?:\\.|[^\"\\])*\"          |
                '(?:\\.|[^'\\])*'             |
                //[^\n]*                      |
                /\*(?:.|\n)*?\*/
            )
            |
            \b({re.escape(old)})\b
            """,
            re.VERBOSE | re.DOTALL,
        )

        def replacer(match: re.Match):
            if match.group(1) is not None:
                return match.group(1)
            return new

        js_files = []
        for ext in ("*.js", "*.jsx", "*.ts", "*.tsx", "*.mjs"):
            js_files.extend(glob.glob(os.path.join(root, "**", ext), recursive=True))

        modified_count = 0
        for filepath in js_files:
            try:
                with open(filepath, "r", encoding="utf-8") as f:
                    source = f.read()
            except OSError:
                continue
            if old in source:
                new_source = pattern.sub(replacer, source)
                if new_source != source:
                    try:
                        with open(filepath, "w", encoding="utf-8") as f:
                            f.write(new_source)
                        modified_count += 1
                    except OSError:
                        pass
        return Count(value=modified_count)

File: lint-arwaky/src-rust/infrastructure/javascript_call_tracer.rs
/// javascript_call_tracer — Semantic analysis adapter for JavaScript/TypeScript files.
use crate::contract::ISemanticTracerPort;
use crate::taxonomy::{
    CallChainList, Count, DataFlowList, DirectoryPath, ExitCode, FilePath, LineNumber, MetadataVO,
    ResponseData, ResponseDataList, ScopeRef, SemanticError, StdError, StdOutput, SymbolName,
    SymbolNameList,
};
use async_trait::async_trait;
use regex::Regex;
use std::fs;
use std::path::{Path, PathBuf};

pub struct JSTracer;

impl JSTracer {
    pub fn new() -> Self {
        Self
    }

    fn get_variant_dict_raw(name: &str) -> std::collections::HashMap<String, String> {
        let words: Vec<String> = Regex::new(r"[A-Za-z][a-z0-9]*|[A-Z]+(?=[A-Z][a-z0-9]|\b)|[0-9]+")
            .unwrap()
            .find_iter(name)
            .map(|m| m.as_str().to_lowercase())
            .collect();

        if words.is_empty() {
            let mut m = std::collections::HashMap::new();
            m.insert("snake_case".to_string(), name.to_string());
            m.insert("camel_case".to_string(), name.to_string());
            m.insert("pascal_case".to_string(), name.to_string());
            m.insert("screaming_snake".to_string(), name.to_uppercase());
            return m;
        }

        let snake = words.join("_");
        let first = words[0].clone();
        let rest: String = words[1..]
            .iter()
            .map(|w| {
                let mut c = w.chars();
                match c.next() {
                    Some(ch) => ch.to_uppercase().to_string() + c.as_str(),
                    None => String::new(),
                }
            })
            .collect();

        let pascal: String = words
            .iter()
            .map(|w| {
                let mut c = w.chars();
                match c.next() {
                    Some(ch) => ch.to_uppercase().to_string() + c.as_str(),
                    None => String::new(),
                }
            })
            .collect();

        let mut m = std::collections::HashMap::new();
        m.insert("snake_case".to_string(), snake.clone());
        m.insert("camel_case".to_string(), format!("{}{}", first, rest));
        m.insert("pascal_case".to_string(), pascal);
        m.insert("screaming_snake".to_string(), snake.to_uppercase());
        m
    }

    fn build_variants_raw(name: &str) -> Vec<String> {
        let d = Self::get_variant_dict_raw(name);
        let kebab = d.get("snake_case").unwrap().replace("_", "-");

        let mut set = std::collections::HashSet::new();
        set.insert(name.to_string());
        set.insert(d.get("snake_case").unwrap().to_string());
        set.insert(d.get("camel_case").unwrap().to_string());
        set.insert(d.get("pascal_case").unwrap().to_string());
        set.insert(d.get("screaming_snake").unwrap().to_string());
        set.insert(kebab);

        set.into_iter().collect()
    }

    fn find_js_files(root: &Path) -> Vec<PathBuf> {
        let mut js_files = Vec::new();
        let mut dirs = vec![root.to_path_buf()];

        while let Some(dir) = dirs.pop() {
            if let Ok(entries) = fs::read_dir(&dir) {
                for entry in entries.flatten() {
                    let path = entry.path();
                    if path.is_dir() {
                        dirs.push(path);
                    } else if path.is_file() {
                        if let Some(ext) = path.extension().and_then(|e| e.to_str()) {
                            if matches!(ext, "js" | "jsx" | "ts" | "tsx" | "mjs") {
                                js_files.push(path);
                            }
                        }
                    }
                }
            }
        }
        js_files
    }
}

#[async_trait]
impl ISemanticTracerPort for JSTracer {
    async fn get_enclosing_scope(
        &self,
        _file_path: &FilePath,
        _line: LineNumber,
    ) -> Result<Option<ScopeRef>, SemanticError> {
        Ok(None)
    }

    async fn trace_call_chain(
        &self,
        root_dir: &DirectoryPath,
        target_name: &SymbolName,
    ) -> Result<CallChainList, SemanticError> {
        let mut callers = Vec::new();
        let name = target_name.to_string();
        let root = Path::new(root_dir.as_str());

        let call_pattern = Regex::new(&format!(r"\b{}\s*\(", regex::escape(&name))).unwrap();
        let def_pattern =
            Regex::new(&format!(r"(?:function|class)\s+{}\b", regex::escape(&name))).unwrap();

        let js_files = Self::find_js_files(root);

        for filepath in js_files {
            if let Ok(content) = fs::read_to_string(&filepath) {
                for (i, line) in content.lines().enumerate() {
                    if call_pattern.is_match(line) && !def_pattern.is_match(line) {
                        if let Ok(rel_path) = filepath.strip_prefix(root) {
                            callers.push(format!(
                                "{}:{} -> {}",
                                rel_path.display(),
                                i + 1,
                                line.trim()
                            ));
                        }
                    }
                }
            }
        }

        Ok(CallChainList { values: callers })
    }

    async fn find_flow(
        &self,
        _file_path: &FilePath,
        _var_name: &SymbolName,
        _start_line: LineNumber,
    ) -> DataFlowList {
        DataFlowList { values: Vec::new() }
    }

    async fn get_variant_dict(&self, name: &SymbolName) -> ResponseData {
        let dict = Self::get_variant_dict_raw(&name.to_string());
        let mut map = std::collections::HashMap::new();
        for (k, v) in dict {
            map.insert(k, serde_json::Value::String(v));
        }
        ResponseData::new(
            StdOutput::new(""),
            StdError::new(""),
            ExitCode::new(0),
            MetadataVO::new(map),
        )
    }

    async fn project_wide_rename(
        &self,
        root_dir: &DirectoryPath,
        old_name: &SymbolName,
        new_name: &SymbolName,
    ) -> u32 {
        let root = Path::new(root_dir.as_str());
        let old = old_name.to_string();
        let new = new_name.to_string();

        let pattern = Regex::new(&format!(
            r"(?x)
            (
                `(?:\\.|[^`\\])*`             |
                \x22(?:\\.|[^\\x22\\])*\x22   |
                '(?:\\.|[^'\\])*'             |
                //[^\n]*                      |
                /\*(?:.|\n)*?\*/
            )
            |
            \b({})\b
            ",
            regex::escape(&old)
        ))
        .unwrap();

        let js_files = Self::find_js_files(root);
        let mut modified_count = 0;

        for filepath in js_files {
            if let Ok(source) = fs::read_to_string(&filepath) {
                if source.contains(&old) {
                    let new_source = pattern.replace_all(&source, |caps: &regex::Captures| {
                        if let Some(m) = caps.get(1) {
                            m.as_str().to_string()
                        } else {
                            new.clone()
                        }
                    });

                    if new_source != source {
                        if fs::write(&filepath, new_source.as_ref()).is_ok() {
                            modified_count += 1;
                        }
                    }
                }
            }
        }

        modified_count
    }

    async fn get_symbol_locations(
        &self,
        _file_path: &FilePath,
        _symbol: &SymbolName,
    ) -> Vec<ResponseData> {
        Vec::new()
    }

    async fn build_variants(&self, name: &SymbolName) -> Vec<SymbolName> {
        Self::build_variants_raw(&name.to_string())
            .into_iter()
            .filter_map(|s| SymbolName::try_from(s).ok())
            .collect()
    }
}
